cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -0,0 +1,309 @@
+                const size_t end   = iBlock + 1 == nBlocks ? nTest : start + blockSize;
+
+                if (voteWeights == VoteWeights::voteUniform)
+                {
+                    s |= uniformWeightedVoting(nClasses, k, end - start, neighborsIndices + start * k, trainLabel, testLabel + start);
+                }
+                else
+                {
+                    s |= distanceWeightedVoting(nClasses, k, end - start, neighborsDistances + start * k, neighborsIndices + start * k, trainLabel,
+                                                testLabel + start);
+                }
+            });
+        }
+
+        return s.detach();
+    }
+
+protected:
+    services::Status uniformWeightedVoting(const size_t nClasses, const size_t k, const size_t n, int * indices, const int * trainLabel,
+                                           int * testLabel)
+    {
+        daal::services::internal::TNArray<int, classBufSize, cpu> classWeightsArr(nClasses);
+        int * classWeights = classWeightsArr.get();
+        DAAL_CHECK_MALLOC(classWeights);
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                classWeights[j] = 0;
+            }
+            for (size_t j = 0; j < k; ++j)
+            {
+                classWeights[trainLabel[indices[i * k + j]]] += 1;
+            }
+            size_t maxWeightClass = 0;
+            size_t maxWeight      = 0;
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                if (classWeights[j] > maxWeight)
+                {
+                    maxWeight      = classWeights[j];
+                    maxWeightClass = j;
+                }
+            }
+            testLabel[i] = maxWeightClass;
+        }
+        return services::Status();
+    }
+
+    services::Status distanceWeightedVoting(const size_t nClasses, const size_t k, const size_t n, FPType * distances, int * indices,
+                                            const int * trainLabel, int * testLabel)
+    {
+        daal::services::internal::TNArray<FPType, classBufSize, cpu> classWeightsArr(nClasses);
+        FPType * classWeights = classWeightsArr.get();
+        DAAL_CHECK_MALLOC(classWeights);
+
+        const FPType epsilon = daal::services::internal::EpsilonVal<FPType>::get();
+        bool isContainZero   = false;
+        for (size_t i = 0; i < k * n; ++i)
+        {
+            if (distances[i] < epsilon && distances[i] > -epsilon)
+            {
+                isContainZero = true;
+                break;
+            }
+        }
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            for (size_t j = 0; j < nClasses; ++j)
+            {
+                classWeights[j] = 0;
+            }
+            for (size_t j = 0; j < k; ++j)
+            {
+                if (isContainZero)
+                {
+                    if (distances[i] < epsilon && distances[i] > -epsilon)
```

Comment:
Why do you need second term (distances[i] > -epsilon) of condition here?

Reply:
In future case for distances with negative values

Reply:
> In future case for distances with negative values

What are distances with negative values?